_drafts/gnu-build-id-for-firmware.md

```diff
@@ -0,0 +1,270 @@
+Idx Name          Size      VMA       LMA       File off  Algn
+  0 .text         000004e4  00000000  00000000  00010000  2**2
+                  CONTENTS, ALLOC, LOAD, READONLY, CODE
+  1 .gnu_build_id 00000024  000004e4  000004e4  000104e4  2**2
+                  CONTENTS, ALLOC, LOAD, READONLY, DATA
+  2 .bss          00000000  20000000  20000000  00000000  2**0
+                  ALLOC
+  3 .data         00000000  20000000  20000000  00010508  2**0
+                  CONTENTS, ALLOC, LOAD, DATA
+  4 .stack        00002000  20000000  20000000  00020000  2**0
+                  ALLOC
+  5 .debug_info   00004076  00000000  00000000  00010508  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  6 .debug_abbrev 00000ac3  00000000  00000000  0001457e  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  7 .debug_aranges 000000f8  00000000  00000000  00015041  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  8 .debug_ranges 000000b8  00000000  00000000  00015139  2**0
+                  CONTENTS, READONLY, DEBUGGING
+  9 .debug_line   00000cfa  00000000  00000000  000151f1  2**0
+                  CONTENTS, READONLY, DEBUGGING
+ 10 .debug_str    0000111a  00000000  00000000  00015eeb  2**0
+                  CONTENTS, READONLY, DEBUGGING
+ 11 .comment      00000075  00000000  00000000  00017005  2**0
+                  CONTENTS, READONLY
+ 12 .ARM.attributes 0000002c  00000000  00000000  0001707a  2**0
+                  CONTENTS, READONLY
+ 13 .debug_frame  00000298  00000000  00000000  000170a8  2**2
+                  CONTENTS, READONLY, DEBUGGING
+```
+
+As you can see, our build ID now has an address assigned to it and is marked
+with the `LOAD` attribute.
+
+**Note**: Make sure to declare the `.gnu_build_id` section after the `.text`
+section, otherwise the build ID will be set to address `0x0` and the firmware
+will not boot.
+
+## Reading the build ID in firmware
+
+Once the build ID is in our binary, we need to modify our firmware to read it
+and, at the very least, print it over serial at boot.
+
+From the linker script, we know that we will find the build ID section at
+`&g_note_build_id`. From the spec, we know the structure of the section and can
+write down a typedef:
+
+```c
+typedef struct {
+    uint32_t namesz;
+    uint32_t descsz;
+    uint32_t type;
+    uint8_t data[];
+} ElfNoteSection_t;
+
+extern const ElfNoteSection_t *g_note_build_id;
+```
+
+We can now simply index into the data field and print the build ID data.
+
+```c
+void print_build_id(void) {
+    const uint8_t *build_id_data = &g_note_build_id->data[g_note_build_id->namesz];
+
+    printf("Build ID: ");
+    for (int i = 0; i < g_note_build_id->descsz; ++i) {
+        printf("%02x", build_id_data[i]);
+    }
+    printf("\n");
+}
+```
+
+## References
+
+- [Original build ID proposal](https://fedoraproject.org/wiki/RolandMcGrath/BuildID)
+- [Elf Note section description](https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-18048.html)
```

Comment:
For fun perhaps also point to the ld sources:
https://github.com/bminor/binutils-gdb/blob/4de283e4b5f21207fe12f99913d1f28d4f07843c/ld/emultempl/elf32.em#L1165

Reply:
sweet, yes!